### ISSUE TITLE
Problem with using custom context menu.

### ISSUE THREAD

**Issue by ajtruckle** [MicrosoftEdge/WebView2Feedback#2368]
This is related to: https://github.com/MicrosoftEdge/WebView2Feedback/issues/2138.

In the first instance I thought I would try to use the existing custom context menu as shown in the documentation and sample.

My code is based on samples. So, I have a core structure:

```
struct CWebBrowserImpl
{
   wil::com_ptr<ICoreWebView2Environment>    m_webViewEnvironment;
   wil::com_ptr<ICoreWebView2Environment2>   m_webViewEnvironment2;
   wil::com_ptr<ICoreWebView2>               m_webView;
   wil::com_ptr<ICoreWebView2_2>             m_webView2;
   wil::com_ptr<ICoreWebView2_11>            m_webView2_11;
   wil::com_ptr<ICoreWebView2Controller>     m_webController;
   wil::com_ptr<ICoreWebView2Controller3>    m_webController3;
   wil::com_ptr<ICoreWebView2Settings>       m_webSettings;
};
```

To my knowledge this exposes the needed interfaces as variables. They are populated here:

```
HRESULT CWebBrowser::OnCreateWebViewControllerCompleted(
   HRESULT result, 
   ICoreWebView2Controller* controller)
{
   if (result == S_OK)
   {
      if (controller != nullptr)
      {
         m_pImpl->m_webController = controller;
         CHECK_FAILURE(controller->get_CoreWebView2(&m_pImpl->m_webView));
		 CHECK_FAILURE(m_pImpl->m_webController->QueryInterface(IID_PPV_ARGS(&m_pImpl->m_webController3)));
		 // m_controller3 = m_controller.try_query<ICoreWebView2Controller3>();

         if (!m_pImpl->m_webView)
            return E_FAIL;

         CHECK_FAILURE(m_pImpl->m_webView->QueryInterface(IID_PPV_ARGS(&m_pImpl->m_webView2)));
		 CHECK_FAILURE(m_pImpl->m_webView->QueryInterface(IID_PPV_ARGS(&m_pImpl->m_webView2_11)));

         CHECK_FAILURE(m_pImpl->m_webView->get_Settings(&m_pImpl->m_webSettings));

         RegisterEventHandlers();

         ResizeToClientArea();
      }

      auto callback = m_callbacks[CallbackType::CreationCompleted];
      if (callback != nullptr)
         RunAsync(callback);
   }
   else
   {
      ShowFailure(result, L"Cannot create webview environment.");
   }

   return S_OK;
}

```

I then have the context menu handler added, like this:

```
void CChristianLifeMinistryEditorDlg::InitialiseWebPreview2()
{
	m_pWebBrowser = std::make_unique<CWebBrowser>();
	if (m_pWebBrowser != nullptr)
	{
		m_pWebBrowser->CreateAsync(
			WS_VISIBLE | WS_CHILD,
			m_rctHtmlPreview,
			this,
			1,
			[this]() {
				m_pWebBrowser->SetParent(this);
				m_pWebBrowser->DisablePopups();
				m_pWebBrowser->DisplayContextMenu();

				UpdatePreview();


			});
	}
}
```

The `DisplayContextMenu` handler is:

```
void CWebBrowser::DisplayContextMenu()
{
    if (m_pImpl->m_webView2_11 != nullptr && 
		m_pImpl->m_webController != nullptr &&
		m_pImpl->m_webController3 != nullptr)
    {
		m_pImpl->m_webView2_11->add_ContextMenuRequested(
			Callback<ICoreWebView2ContextMenuRequestedEventHandler>(
				[this](
					ICoreWebView2* sender,
					ICoreWebView2ContextMenuRequestedEventArgs* eventArgs) {
						wil::com_ptr<ICoreWebView2ContextMenuRequestedEventArgs> args =
							eventArgs;
						wil::com_ptr<ICoreWebView2ContextMenuItemCollection> items;
						CHECK_FAILURE(args->get_MenuItems(&items));
						UINT32 itemsCount;
						CHECK_FAILURE(items->get_Count(&itemsCount));

						wil::com_ptr<ICoreWebView2ContextMenuTarget> target;
						CHECK_FAILURE(args->get_ContextMenuTarget(&target));

						COREWEBVIEW2_CONTEXT_MENU_TARGET_KIND targetKind;
						CHECK_FAILURE(target->get_Kind(&targetKind));
						// Custom UI Context Menu rendered if invoked on selected text
						if (targetKind ==
							COREWEBVIEW2_CONTEXT_MENU_TARGET_KIND_SELECTED_TEXT)
						{
							auto showMenu = [this, args, itemsCount, items, target] {
								CHECK_FAILURE(args->put_Handled(true));
								HMENU hPopupMenu = CreatePopupMenu();
								AddMenuItems(hPopupMenu, items);
								HWND hWnd;
								m_pImpl->m_webController->get_ParentWindow(&hWnd);
								::SetForegroundWindow(hWnd);
								RECT rct;
								::GetClientRect(hWnd, &rct);
								POINT topLeft;
								topLeft.x = rct.left;
								topLeft.y = rct.top;
								::ClientToScreen(hWnd, &topLeft);
								POINT p;
								CHECK_FAILURE(args->get_Location(&p));
								RECT bounds;
								CHECK_FAILURE(m_pImpl->m_webController->get_Bounds(&bounds));
								double scale;
								m_pImpl->m_webController3->get_RasterizationScale(&scale);
								INT32 selectedCommandId = ::TrackPopupMenu(
									hPopupMenu,
									TPM_TOPALIGN | TPM_LEFTALIGN | TPM_RETURNCMD,
									bounds.left + topLeft.x + ((int)(p.x * scale)),
									bounds.top + topLeft.y + ((int)(p.y * scale)), 0, hWnd,
									NULL);
								CHECK_FAILURE(
									args->put_SelectedCommandId(selectedCommandId));
							};
							wil::com_ptr<ICoreWebView2Deferral> deferral;
							CHECK_FAILURE(args->GetDeferral(&deferral));
							RunAsync([deferral, showMenu]() {
								showMenu();
								CHECK_FAILURE(deferral->Complete());
								});
							//m_appWindow->RunAsync([deferral, showMenu]() {
							//	showMenu();
							//	CHECK_FAILURE(deferral->Complete());
							//	});
						}
						// Removing the 'Save image as' context menu item for image context
						// selections.
						else if (targetKind == COREWEBVIEW2_CONTEXT_MENU_TARGET_KIND_IMAGE)
						{
							UINT32 removeIndex = itemsCount;
							wil::com_ptr<ICoreWebView2ContextMenuItem> current;
							for (UINT32 i = 0; i < itemsCount; i++)
							{
								CHECK_FAILURE(items->GetValueAtIndex(i, &current));
								wil::unique_cotaskmem_string name;
								CHECK_FAILURE(current->get_Name(&name));
								if (std::wstring(L"saveImageAs") == name.get())
								{
									removeIndex = i;
								}
							}
							if (removeIndex < itemsCount)
							{
								CHECK_FAILURE(items->RemoveValueAtIndex(removeIndex));
							}
						}
						// Adding a custom context menu item for the page that will display
						// the page's URI.
						else if (targetKind == COREWEBVIEW2_CONTEXT_MENU_TARGET_KIND_PAGE)
						{
							// Custom items should be reused whenever possible.
							if (!m_displayPageUrlContextSubMenuItem)
							{
								wil::com_ptr<ICoreWebView2Environment9> webviewEnvironment;
								CHECK_FAILURE(
									m_pImpl->m_webViewEnvironment->QueryInterface(
										IID_PPV_ARGS(&webviewEnvironment)));
								//CHECK_FAILURE(
								//	m_appWindow->GetWebViewEnvironment()->QueryInterface(
								//		IID_PPV_ARGS(&webviewEnvironment)));
								wil::com_ptr<ICoreWebView2ContextMenuItem>
									displayPageUrlItem;
								wil::com_ptr<IStream> iconStream;
								CHECK_FAILURE(SHCreateStreamOnFileEx(
									L"small.ico", STGM_READ, FILE_ATTRIBUTE_NORMAL, FALSE,
									nullptr, &iconStream));
								CHECK_FAILURE(webviewEnvironment->CreateContextMenuItem(
									L"Display Page Url", iconStream.get(),
									COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND_COMMAND,
									&displayPageUrlItem));
								CHECK_FAILURE(displayPageUrlItem->add_CustomItemSelected(
									Callback<ICoreWebView2CustomItemSelectedEventHandler>(
										[appWindow = this, target](
											ICoreWebView2ContextMenuItem* sender,
											IUnknown* args) {
												wil::unique_cotaskmem_string pageUri;
												CHECK_FAILURE(target->get_PageUri(&pageUri));
												appWindow->AsyncMessageBox(
													pageUri.get(), L"Display Page Uri");
												return S_OK;
										})
									.Get(),
											nullptr));
								CHECK_FAILURE(webviewEnvironment->CreateContextMenuItem(
									L"New Submenu", nullptr,
									COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND_SUBMENU,
									&m_displayPageUrlContextSubMenuItem));
								wil::com_ptr<ICoreWebView2ContextMenuItemCollection>
									m_displayPageUrlContextSubMenuItemChildren;
								CHECK_FAILURE(
									m_displayPageUrlContextSubMenuItem->get_Children(
										&m_displayPageUrlContextSubMenuItemChildren));
								m_displayPageUrlContextSubMenuItemChildren
									->InsertValueAtIndex(0, displayPageUrlItem.get());
							}
							CHECK_FAILURE(items->InsertValueAtIndex(
								itemsCount, m_displayPageUrlContextSubMenuItem.get()));
						}
						// If type is not an image, video, or page, the regular Edge context
						// menu will be displayed
						return S_OK;
				})
			.Get(),
					&m_contextMenuRequestedToken);

		::MessageBox(
			nullptr, L"Custom Context menus are now enabled.", L"Settings change",
			MB_OK);

    }
}

```
I am confused because when I run my code it does show in a popup _Custom Context menus are now enabled._. But then, when I try to right click the control I get an exception:

![Exception](https://user-images.githubusercontent.com/7985331/163722526-5f969caf-5f8c-4646-9036-fa60c11aaf20.png)

The `CTTComboBox` class is used for a control on the dialog.

What have I done wrong? I can't share my complete solution but I can attach the WebView related classes. [](url
[WebView2Classes.zip](https://github.com/MicrosoftEdge/WebView2Feedback/files/8501955/WebView2Classes.zip)).

Any clarification on my mistakes is appreciated.  I feel it must be much simpler than this to support my own context menu (which is a mixture of Edge controls and my own controls). Just wanted to start with sample code.

**Comment by ajtruckle** [comment]
For your info, this is the [`CTTComboBox`](https://www.codeproject.com/articles/10559/neat-tooltip-for-combobox) class. I have three of these combo controls on my dialog. I am unable to proceed with testing due to either a bug on my part or a conflict between the control class and the web control.

**Comment by ajtruckle** [comment]
I have made **some** progress with this. I had a closer look at the _call stack_ and determined that I did not have the **small.ico** file. I placed it on my D drive and changed the code:

```
CHECK_FAILURE(SHCreateStreamOnFileEx(
	L"d:\\small.ico", STGM_READ, FILE_ATTRIBUTE_NORMAL, FALSE,
	nullptr, &iconStream));
```

Now it does not crash when I right-click and I see the menu:

![ContextMenu](https://user-images.githubusercontent.com/7985331/163838437-8507bece-af5a-4057-b02b-9296d5cd34b7.png)

The only issue is that the icon is not visible. It is visible in the sample application.
